File: 원종/16주차_2229.c
#include <stdio.h>

#define ABS(x)		( ((x)<0)?-(x):(x) )
#define MAX(a, b)	a > b ? a : b

int arr[1000];
int d[1001];

int main() {

	int n;
	scanf("%d", &n);
	for (int i = 0; i < n; i++)
		scanf("%d", &arr[i]);

	d[1] = ABS(arr[1] - arr[0]);
	for (int i = 2; i < n; i++) 
		for (int j = 1; j < i; j++) {
			d[i] = MAX(d[i], ABS(arr[j] - arr[i]) + d[j - 1]);
			d[i] = MAX(d[i], d[i - 1]);
		}
	printf("%d\n", d[n - 1]);
	return 0;
}

File: 은찬/16주차_2229.java
import java.io.*;
import java.util.*;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());

        int[] arr = new int[n];
        int[] dp = new int[n];
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }

        for (int y = 1; y < n; y++) {
            for (int x = 0; x <= y; x++) {
                int abs = Math.abs(arr[y] - arr[x]);
                int next;
                if (x == 0) next = abs;
                else next = abs + dp[x - 1];
                dp[y] = Math.max(dp[y], next);
            }
        }

        bw.write(String.valueOf(dp[n - 1]));

        br.close();
        bw.close();
    }
}